clients/src/main/java/org/apache/kafka/common/security/ssl/SslFactory.java

```diff
@@ -101,88 +101,118 @@ public void configure(Map<String, ?> configs) throws KafkaException {
                         "can't connect to a server SSLEngine created with those settings.", e);
             }
         }
-        this.sslEngineBuilder = builder;
+        this.sslEngineFactory = builder;
     }
 
     @Override
     public Set<String> reconfigurableConfigs() {
-        return SslConfigs.RECONFIGURABLE_CONFIGS;
+        return sslEngineFactory.reconfigurableConfigs();
     }
 
     @Override
     public void validateReconfiguration(Map<String, ?> newConfigs) {
-        createNewSslEngineBuilder(newConfigs);
+        createNewSslEngineFactory(newConfigs);
     }
 
     @Override
     public void reconfigure(Map<String, ?> newConfigs) throws KafkaException {
-        SslEngineBuilder newSslEngineBuilder = createNewSslEngineBuilder(newConfigs);
-        if (newSslEngineBuilder != this.sslEngineBuilder) {
-            this.sslEngineBuilder = newSslEngineBuilder;
+        SslEngineFactory newSslEngineFactory = createNewSslEngineFactory(newConfigs);
+        if (newSslEngineFactory != this.sslEngineFactory) {
+            this.sslEngineFactory = newSslEngineFactory;
             log.info("Created new {} SSL engine builder with keystore {} truststore {}", mode,
-                    newSslEngineBuilder.keystore(), newSslEngineBuilder.truststore());
+                    newSslEngineFactory.keystore(), newSslEngineFactory.truststore());
         }
     }
 
-    private SslEngineBuilder createNewSslEngineBuilder(Map<String, ?> newConfigs) {
-        if (sslEngineBuilder == null) {
+    private SslEngineFactory instantiateSslEngineFactory(Map<String, Object> configs) {
+        String sslEngineFactoryClass = (String) configs.getOrDefault(SslConfigs.SSL_ENGINE_FACTORY_CLASS_CONFIG,
+                DefaultSslEngineFactory.class.getCanonicalName());
+        try {
+            SslEngineFactory sslEngineFactory = Utils.newInstance(sslEngineFactoryClass, SslEngineFactory.class);
+            sslEngineFactory.configure(configs);
+            this.sslEngineFactoryConfig = configs;
+            return sslEngineFactory;
+        } catch (ClassNotFoundException e) {
```

Comment:
We would have failed earlier once we make the factory type a CLASS instead of STRING.